### ISSUE TITLE
Failure to train distributed with strategy mirrored, custom training, functions supplied in class

### ISSUE THREAD

**Issue by ThWuensche** [tensorflow/tensorflow#46373]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Debian Buster with ROCm 3.7
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): tensorflow-rocm 2.3.0 -  v2.3.0-rc1-2358-gc0826c7973 2.3.0
- Python version: 3.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory: Radeon VII

**Describe the current behavior**

Training works on single GPU, fails on dual GPU

**Describe the expected behavior**

Should also work on dual GPU

**Standalone code to reproduce the issue**

I'm trying to run YOLOV3 tensorflow from https://github.com/ethanyanjiali/deep-vision.git with VOC_2007 data set

**Other info / logs**

I compared the code to example code for MirroredStrategy and could not find any ad hoc visible mistake. Only major difference is that the training is realized in a separate trainer class.

By running in standard eager execution I get the following error output: [log.txt](https://github.com/tensorflow/tensorflow/files/5804098/log.txt)

When switching off eager execution I get the following output (which might indicate the problem, if the change from eager to non-eager does not create systematic difference): [log_no_eager.txt](https://github.com/tensorflow/tensorflow/files/5804099/log_no_eager.txt)

I moved the generation of the optimizer into the mirrored strategy scope in main and passed it as parameter to the constructor of the trainer class, but that did not solve the problem.

The error message seems similar to [that issue](https://github.com/tensorflow/tensorflow/issues/41768), but there was no reproducible solution mentioned.

**Comment by amahendrakar** [comment]
> Training works on single GPU, fails on dual GPU

@ThWuensche,
Could you please provide the make and model of both the GPUs on your machine. 

Also, please run the below code snippet and share the output with us. 
```
import tensorflow as tf
print("Num GPUs Available: ", len(tf.config.experimental.list_physical_devices('GPU')))
```
Thanks!

**Comment by ThWuensche** [comment]
As mentioned the GPUs are AMD Radeon VII, running with ROCm 3.7.0, tensorflow-rocm 2.3.0

Here is the output from running the code:
>>> import tensorflow as tf
>>> print("Num GPUs Available: ", len(tf.config.experimental.list_physical_devices('GPU')))
2021-01-13 19:03:44.609744: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libamdhip64.so
2021-01-13 19:03:48.603715: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1734] Found device 0 with properties: 
pciBusID: 0000:0a:00.0 name: Device 66af     ROCm AMD GPU ISA: gfx906
coreClock: 1.801GHz coreCount: 60 deviceMemorySize: 15.98GiB deviceMemoryBandwidth: 0B/s
2021-01-13 19:03:48.603802: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1734] Found device 1 with properties: 
pciBusID: 0000:0d:00.0 name: Device 66af     ROCm AMD GPU ISA: gfx906
coreClock: 1.801GHz coreCount: 60 deviceMemorySize: 15.98GiB deviceMemoryBandwidth: 0B/s
2021-01-13 19:03:48.650101: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library librocblas.so
2021-01-13 19:03:48.652980: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libMIOpen.so
2021-01-13 19:03:48.673788: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library librocfft.so
2021-01-13 19:03:48.675584: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library librocrand.so
2021-01-13 19:03:48.675730: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1858] Adding visible gpu devices: 0, 1
Num GPUs Available:  2
>>>

**Comment by ThWuensche** [comment]
I have a second machine with four of these GPUs, there the training should finally run

**Comment by amahendrakar** [comment]
@ThWuensche,
Could you please update TensorFlow to the latest stable version v2.4 and check if you are facing the same issue. Thanks!

**Comment by ThWuensche** [comment]
The issue is not solved yet. Unfortunately the feedback in form of e-mail reply to the notifications did not make it into the issue communication.

Could the issue please be reopened, as far as I understand I can not do that if it was closed by a repo maintainer?

I have in the meantime installed rocm tensorflow 2.4 in a docker container (docker pull rocm/tensorflow:rocm4.0.1-tf2.4-dev from hub.docker.com), but the problem persists.

**Comment by ThWuensche** [comment]
Looks like the code triggers some problems in tensorflow or the rocm implementation. With the above mentioned version of tensorflow it works in single GPU configuration, but if I run the same code in rocm/tensorflow:latest, the validate step delivers NaNs.

Multi-GPU works in none of the two versions.

It's not my code, but I couldn't spot any reasons for problems. Would be interesting to see whether it behaves differently on NVidia, however I have only AMD GPUs.

**Comment by ThWuensche** [comment]
Thanks for reopening! Is there anything I can do to help analyze and fix the issue? The YOLO package used is easy to install, but needs an extra step to prepare the dataset. Should I prepare a tarball of the dataset?

**Comment by nikitamaia** [comment]
Hi @ThWuensche, in #41768 the problem was the use of `tf.compat.v1.disable_eager_execution()`. Disabling eager execution is not recommended in TF2, and can cause issues when using a distribution strategy.

However, since you're facing problems even when eager execution is not disabled, it seems that we need to do some more investigation to figure out what's going on here.
- Can you share what happens if you replace MirroredStrategy with `tf.distribute.get_strategy()`?
- This is a lot of code for us to parse through. Can you try and trim it down (ideally something we can even run in colab)?
- Did you write this code? If not I think this issue is probably better filed in the repo you got the code from.

**Comment by ThWuensche** [comment]
Hi @nikitamaia, thanks for your reply. I just returned from a trip and have to get things set up again, I will try your proposal.

I did not write that code, but used and modified it for a plant detection system for agriculture. So I would be interested to get that solved. Sure it's a lot of code, but I had done investigation before opening the issue and think the code pretty much resembles the examples for distributed execution, the major difference being that in this code it is packed in clesses, while in the examples no encapsulation in classes is used. Maybe that changes some memory layout?

Basically I think if an application works non-distributed, it should also work distributed, having the distribution handled transparently by tensorflow. If some property of the application prevents that, ideally tensorflow should point out what's the problem and how it can be fixed. That's why I opened the issue here. So far I don't see any indication it's the applications fault.

I'm running with ROCm on Radeon VIIs, don't have NVidia GPUs to test.

**Comment by nikitamaia** [comment]
Hi, I definitely agree that if an application works in the non-distributed case, it should also work in a distributed environment! However, I am not sure what code you running. The link you provided in your original post is for a repo with many models in it, and doesn't seem to have the distributed code in it at all. Without that it's very difficult for us to diagnose and determine if there was an issue when setting up the distribution strategy, or if the error is something in the tf.distribute.Strategy API. We get lots and lots of issues filed here, so the more narrowed down the code is, the quicker we can help figure out the problem. Thanks!

**Comment by ThWuensche** [comment]
Dear @nikitamaia, I'm running the YOLOV3 model in that repository.

The relevant code seems to be in "run" method in:
https://github.com/ethanyanjiali/deep-vision/blob/master/YOLO/tensorflow/train.py

There are two tf.functions:

At line 125:
        @tf.function
        def distributed_train_epoch(dataset, train_summary_writer,
                                    total_steps):
 
and at line 182:
        @tf.function
        def distributed_val_epoch(dataset):

Hope that information is helpful in nailing down the problem.

Only thing I had to change to run the code is "self.strategy.experimental_run_v2" to just "self.strategy.run"

**Comment by nikitamaia** [comment]
Sorry for the bot trying to close this issue! I forgot to remove the awaiting response label :)

To clarify: [you ran the code in this file](https://github.com/ethanyanjiali/deep-vision/blob/master/YOLO/tensorflow/train.py) as is with no changes, except for changing `experimental_run_v2` to `run`? Also, I'm assuming [you followed the instructions here](https://github.com/ethanyanjiali/deep-vision/tree/master/YOLO/tensorflow) to download the data and run the training script? If so my suggestion would be to also file an issue with the person who wrote this code since it seems they have a tutorial that is not working.

It would also be useful to understand what happens if you replace MirroredStrategy with `tf.distribute.get_strategy()`.

**Comment by ThWuensche** [comment]
Dear Nikita,

yes, I ran that code without further changes (besides the mentioned one). As mentioned in the beginning of the issue I installed and worked with the VOC 2007 dataset (not MSCOCO). The dataset installed for me without big problems according to the instructions, only thing I had to provide was "ray".

I wanted to test again on the system I have at hand now with tf.distribute.get_strategy(), but in the moment have hassles to get tf 2.4 working on that system. I remember when I switched to tf 2.4 according to your colleagues proposal, I had to rework most of the basic install of my (other) system. The dependencies AMD is creating with its tensorflow-rocm port are not very nice, close constraints regarding as well python as ROCm version. Most of the time at least one of that constraints is hard to fulfill on Debian standard system setup. So I need to spend some time to create a setup to run the test.

Just to give some feedback to highlight I'm still very interested in the solution.

Best regards, Thomas

**Comment by ThWuensche** [comment]
Dear Nikita,

finally I got around to try with `tf.distribute.get_strategy()`. It throws no error, but I seem to have load only on the second GPU, the first seems to be idle. So looks like it would use only one GPU in that setup, though it lists both as active in the output.

That is the output:
[tf_log_deepvision_getstrategy.txt](https://github.com/tensorflow/tensorflow/files/6579745/tf_log_deepvision_getstrategy.txt)

If I run with MirroredStrategy, I get some warning about nccl and it breaks:
[tf_log_deepvision_unchanged.txt](https://github.com/tensorflow/tensorflow/files/6579863/tf_log_deepvision_unchanged.txt)

The NCCL error seems like a system/rocm issue, will look on that issue further. The behavior seems changed. I was running this time with ROCM 4.0.1.

**Comment by sanjoy** [comment]
@deven-amd any ideas what could be going wrong?

**Comment by sanatmpa1** [comment]
@ThWuensche,

Can you try updating TF to latest stable version i.e 2.7.0 and let us know if the issue still persists in newer version? Thanks!

**Comment by VetleWoie** [comment]
Hi, I know this is marked as closed, but I seem to be experiencing the same issue.

System information

Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
OS Platform and Distribution (e.g., Linux Ubuntu 16.04): SUSE Linux Enterprise Server 15 SP3 with ROCM 5.0.2
Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
TensorFlow installed from (source or binary): binary
TensorFlow version (use command below): tensorflow-rocm  v2.11.0-3797-gfe65ef3bbcf 2.11.0
Python version: 3.9.12
Bazel version (if compiling from source):
GCC/Compiler version (if compiling from source):
CUDA/cuDNN version:
GPU model and memory:  AMD Instinct MI250X 128GB HBM2e

I'm running a small example just to test distribution strategies in TensorFlow. 

The following is the code I'm trying to run.
```python
import os
import sys
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
os.environ['NCCL_DEBUG'] = 'WARN' 
import tensorflow as tf
import pathlib


print("###########################")
print("Python version: ", sys.version)
print("Tensorflow version: ", tf.version.GIT_VERSION, tf.version.VERSION)
print("Physical devices found: ",tf.config.list_physical_devices())
#Choose mirrored strategy
strategy = tf.distribute.MirroredStrategy()
print("Strategy found number of devices: ", strategy.num_replicas_in_sync)
print("###########################")



datadir = pathlib.Path("./data/")
# Preprocessing
train_ds = tf.keras.utils.image_dataset_from_directory(
    directory=datadir,
    batch_size=32,
    image_size=(300,300),
    seed=123,
    validation_split=0
)

AUTOTUNE = tf.data.AUTOTUNE

train_ds = train_ds.cache().prefetch(buffer_size=AUTOTUNE)

num_classes = 2

with strategy.scope():
    model = tf.keras.Sequential([
        #Rescale images before feeding them to the network
        tf.keras.layers.Rescaling(1./255),
        #create a small convolutional network
        tf.keras.layers.Conv2D(32, 3, activation='relu'),
        tf.keras.layers.MaxPooling2D(),
        tf.keras.layers.Conv2D(32, 3, activation='relu'),
        tf.keras.layers.MaxPooling2D(),
        tf.keras.layers.Conv2D(32, 3, activation='relu'),
        tf.keras.layers.MaxPooling2D(),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.Dense(num_classes)
    ])

    model.compile(
        optimizer='adam',
        loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    )

model.fit(
    train_ds,
    epochs=50
)
```
It works if I try not to run with a distribution strategy, it also runs if I use ```tf.distribute.get_strategy()```. But then only on one GPU. 

When running with ```strategy = tf.distribute.MirroredStrategy()``` it gets to the training part of the code.
I know this because it prints this:
```
Epoch 1/50
RCCL version 2.11.4+hip5.1
```
That is how far it gets before it prints a gigantic traceback, at then end this is printed:
```
3 root error(s) found.
  (0) INTERNAL:  NCCL: unhandled system error. Set NCCL_DEBUG=WARN for detail.
	 [[{{node NcclAllReduce}}]]
	 [[update_0_10/AssignAddVariableOp/_91]]
  (1) INTERNAL:  NCCL: unhandled system error. Set NCCL_DEBUG=WARN for detail.
	 [[{{node NcclAllReduce}}]]
	 [[div_no_nan/AddN/_69]]
  (2) INTERNAL:  NCCL: unhandled system error. Set NCCL_DEBUG=WARN for detail.
	 [[{{node NcclAllReduce}}]]
0 successful operations.
```
I then sat ```NCCL_DEBUG=WARN``` to see if I got any more information. and in between the traceback I find this:
```
nid005001:44042:44502 [0] /long_pathname_so_that_rpms_can_package_the_debug_info/data/driver/rccl/src/include/socket.h:415 NCCL WARN Net : Connect to 10.120.24.11<51347> failed : No route to host

nid005001:44042:44503 [1] /long_pathname_so_that_rpms_can_package_the_debug_info/data/driver/rccl/src/include/socket.h:415 NCCL WARN Net : Connect to 10.120.24.11<51347> failed : No route to host
```
